src/webapi/events/drag.rs

```diff
@@ -0,0 +1,623 @@
+        js!(@(no_return) @{self.as_ref()}.remove(@{index}));
+    }
+
+    /// Removes all of the drag items from the list.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/clear)
+    pub fn clear( &self ) {
+        js!(@(no_return) @{self.as_ref()}.clear());
+    }
+
+    /// Getter that returns a DataTransferItem at the given index.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList/DataTransferItem)
+    pub fn index( &self, index: u32 ) -> Option<DataTransferItem> {
+        let v: Value = js!(
+            return @{self.as_ref()}[@{index}];
+        );
+        match v {
+            Value::Reference(_) => Some(v.try_into().unwrap()),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum DataTransferItemAdd {
+    File(File),
+    String(String, String),
+}
+
+/// The DataTransferItem object represents one drag data item. During a drag operation, each drag
+/// event has a dataTransfer property which contains a list of drag data items. Each item in the
+/// list is a DataTransferItem object.
+///
+/// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem)
+#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
+#[reference(instance_of = "DataTransferItem")]
+pub struct DataTransferItem( Reference );
+
+impl DataTransferItem {
+    /// The kind of drag data item, string or file.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/kind)
+    pub fn kind( &self ) -> DataTransferItemKind {
+        let kind: String = js!(
+            return @{self.as_ref()}.kind;
+        ).try_into().unwrap();
+        match kind.as_ref() {
+            "string" => DataTransferItemKind::String,
+            "file" => DataTransferItemKind::File,
+            other => panic!("Expected valid kind, got {:?}", other),
+        }
+    }
+
+    /// The drag data item's type, typically a MIME type.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/type)
+    pub fn ty( &self ) -> String {
+        js!(
+            return @{self.as_ref()}.type;
+        ).try_into().unwrap()
+    }
+
+    /// Returns the File object associated with the drag data item
+    /// (or null if the drag item is not a file)
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/getAsFile)
+    pub fn get_as_file( &self ) -> File {
+        js!(
+            return @{self.as_ref()}.getAsFile();
+        ).try_into().unwrap()
+    }
+
+    /// Invokes the specified callback with the drag data item string as its argument.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/getAsString)
+    #[cfg(feature = "futures-support")]
+    pub fn get_as_string( &self ) -> oneshot::Receiver<String> {
```

Comment:
I'm not against having a future-y method, but it probably should be the other way around. (Have the callback version be called `get_as_string`, and the futures version something else, since the callback one is the original.)